### PR TITLE
Improve plaintext signature management

### PR DESCRIPTION
**Pull request by jan-kiszka**

### CODE REVIEW

**Comment by 4ch1m** [comment]
Nice work, Jan. :+1: 
Could you please point the PR to the [dev](/4ch1m/mozext/tree/dev_signatureswitch-me/) branch?
Also I don't generally want to include CC to the autoswitch. (A dedicated user-option for that would be better imho. But... I will take care of that myself.)
Hope I'll find a free time-slot tomorrow, so I can test this out.
Thanks!

**Comment by jan-kiszka** [comment]
Dropped the CC patch. Looking forward to have that as option in the next release.

**Comment by 4ch1m** [comment]
Hey Jan,
I'm testing your changes right now.
Generally I like the approach of manipulating the DOM directly (instead of replacing the whole body). :+1: 
I might implement something similar for the HTML-case of the code.

Now the bad news... :wink: 
Everyhting works fine for replies now.
However, something is broken when you compose a fresh/new mail.
Just enter some text... and then switch to another signature.
The signature gets replaced but all text above also gets deleted! ... ending up in an empty mail with just the signature.
Still trying to figure out the problem here.
Can you reproduce this on your side?

**Comment by jan-kiszka** [comment]
Yes, confirmed. Precondition seems to be that the signature is added to an empty window and then switched. Let me dig...

**Comment by jan-kiszka** [comment]
OK, filling the signature into an empty window makes the whole content the signature. I suppose some extra <br> in front of it will help.

**Comment by jan-kiszka** [comment]
Understanding was simple, finding a better algorithm again not. But this one works fine so far.

It could possibly further improved by extracting the current plaintext signature from the DOM tree, rather than via the details interface.

**Comment by 4ch1m** [comment]
All right. No worries.

I'm still a bit hesitant to merge this PR though.
IMHO a bit too much fiddling around with the (signature-)body; like forced removal of trailing new-lines:
```
                    while (plainTextBody.endsWith("\n")) {
                        plainTextBody = plainTextBody.substring(0, plainTextBody.length - 1);
                    }
                    let plainTextSignature = await createPlainTextSignature(signature.text);
                    while (plainTextSignature.endsWith("\n")) {
                        plainTextSignature = plainTextSignature.substring(0, plainTextSignature.length - 1);
                    }
```

Also... always adding a new line to every signature at the end (just for the sake of better detection later on?!) seems weird and inconsistent:
```
    return  PLAINTEXT_SIGNATURE_SEPARATOR +
            text +
            (!text.endsWith(NEW_LINE) ? NEW_LINE : "");
```
Any way... I'll try to utilize your work/approach and try to come up with a similar solution.

Stay tuned. :wink:

**Comment by jan-kiszka** [comment]
That adding is needed for TB to not collapse the space between body and signature on conversion to plaintext when sending or saving. That beast is fragile.

And that newline removal above is just for the sake of comparing, not for appending.

**Comment by 4ch1m** [comment]
All right.
I've spent a couple of hours refactoring everything. -> [dev_signatureswitch-me](https://github.com/4ch1m/mozext/tree/dev_signatureswitch-me)
A "compose script" definitely is the way to go.
Thanks for pointing me into that direction, @jan-kiszka !
Much appreciated.

Both plaintext- and html-signatures now get inserted/removed via a compose-script.
This approach also makes a lot other wonky code passages obsolete now; improving general stability. :smile: 

I've published out a beta-release; available here: [2.0.4beta1](http://mozext.achimonline.de/devbuilds.php)

If no regression shows up, I'll push this to master soon.

So... closing this PR now.

(If you like/allow me, I'd put you on my [contributors](http://mozext.achimonline.de/contributors.php)-page. Just let me know.)

**Comment by jan-kiszka** [comment]
There is still something broken, also with my original PR:

```
...
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Language: en-US
Content-Transfer-Encoding: 7bit


First line.
Second line.
```

This is what happens when creating and saving a new plaintext message as draft with the addon active - even if no signature is appended. This is what should happen:

```
...
MIME-Version: 1.0
Content-Type: text/plain; charset=utf-8
Content-Language: en-US
Content-Transfer-Encoding: 7bit

First line.
Second line.
```

I.e. only a single blank line between header and body. I'm also seeing messages that collapse the blank line between body and signature once saved (or sent). This particularly happens when replying to messages.

```
...
MIME-Version: 1.0
In-Reply-To: <0e22186c-7618-3398-05f2-859bbedb752a@web.de>
Content-Type: text/plain; charset=utf-8
Content-Language: en-US
Content-Transfer-Encoding: 7bit


On 14.11.20 12:53, Jan Kiszka wrote:
> Citation
> 
...
> --
> 2.26.2
> 
> 

Last line.
-- 
The Signature

```

As that signature is not in the expected format, reopening that draft for editing will add another signature.

I've compared the DOM trees in both cases (against signatures added by TB natively), but there are no apparent differences. Maybe TB has some additional state that gets out of sync when editing the DOM tree.

**Comment by 4ch1m** [comment]
Thanks, @jan-kiszka .
I'll look into this.

**Comment by 4ch1m** [comment]
Had to invest several hours again to get this fixed. :disappointed: 
Your description pretty much nails it: This really is a fragile beast!

But I think we're good now.
New beta release available here: [2.4.0beta2](http://mozext.achimonline.de/devbuilds.php)

@jan-kiszka : Could you give it another try, please?!

Thanks!

**Comment by jan-kiszka** [comment]
Better! But it still drops the blank line before the signature here when replying to a message, although it looks differently before sending or saving.

**Comment by 4ch1m** [comment]
> But it still drops the blank line before the signature here when replying to a message...

Ha!
Just found out, that this is a "bug" with Thunderbird itself. :smile: 
I fired up a fresh installation of TB (without any add-ons/extensions, etc.) and ran my tests.

Steps to reproduce:
1. create a signature for your account
image
---

2. open a new plaintext-composer and enter a new message
image
(Notice the manually entered new line **BEFORE** the signature.)
---

3. save the message as a draft
---

4. now go to the "Drafts"-folder; select the draft and click <kbd>Reply</kdb>
---

5. the composer opens up again with this ...
image
(Notice that the extra new line from the initial compose remained (see angle bracket), **PLUS** another new line **BEFORE** the signature gets inserted by Thunderbird. You can also position your cursor into that additional new line.)
---

6. now simply copy/paste all of the composer's content (Ctrl+A , Ctrl+C) and paste it somewhere; you'll get this ...
```


On 11/17/20 5:46 PM, Moe Zilla wrote:
> this
> is
> a
> test.
> 
-- 
this is my signature

```
That additional new line before the signature is not carried over / transferred to the clipboard.

You'll get the exact same behaviour with or without SignatureSwitch.

Is that what you've seen/experienced, @jan-kiszka ?

**Comment by jan-kiszka** [comment]
I'm not surprised it's a TB bug. I was just not triggered for me without the addon in normal workflows so far. Filed a bug already?

BTW, I just found a new oddity of the addon: Now that existing signatures get detected in the text, you also replace them if they do not match any of those defined in the addon. Steps: Paste in a git patch, switch on signature, and the git "signature" gets replaced.

```
...
-- 
2.26.2
```

Weren't you matching on the content so far, at least when there was no clear footer div?

**Comment by 4ch1m** [comment]
> I'm not surprised it's a TB bug.

Well... based on your report I was under the assumption that this behavious isn't "normal" (meaning it's not TB's standard behaviour).
But it it clearly is.

> I was just not triggered for me without the addon in normal workflows so far.

Strange.

> Filed a bug already?

No. I'm not sure if I would declare that a "bug" or even an issue.
It sure is unexpected though.

> Weren't you matching on the content so far, at least when there was no clear footer div?

Yep. That was my initial design.
My intention was to only remove what has been inserted by my add-on.
(Which also led to problems when I was trying to detect the random fortune-cookies content.)
I figured out some time ago that many users are using both TB's native signature functionality AND Signature Switch at the same time.
But with the compose-script in place I've decided to identify *any* signature present and remove/replace it.
Also... I was forced to match against signature-content in plaintext-mode in the previous design (in order to find out the id for cycling through the signatures).
But now I also add a custom-attribute (with the sig-id) in plaintext mode.
TLDR; yes... I used to match signature content... but I simplified it now. :)
